pkg/controller/extension/secret.go

```diff
@@ -0,0 +1,131 @@
+package extension
+
+import (
+	"context"
+	"fmt"
+	"maps"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	v1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	fluxv1alpha1 "github.com/stackitcloud/gardener-extension-shoot-flux/pkg/apis/flux/v1alpha1"
+)
+
+const (
+	managedByLabelKey   = "app.kubernetes.io/managed-by"
+	managedByLabelValue = "gardener-extension-" + fluxv1alpha1.ExtensionType
+)
+
+// ReconcileSecrets copies all secrets referenced in the extension (additionalSecretResources or
+// source.SecretResourceName), and deletes all secrets that are no longer referenced.
+// We cannot use gardener resource manager here, because we want to work in the namespace
+// "flux-system", which the resource manager is not configured for.
+func ReconcileSecrets(
+	ctx context.Context,
+	log logr.Logger,
+	seedClient client.Client,
+	shootClient client.Client,
+	seedNamespace string,
+	config *fluxv1alpha1.FluxConfig,
+	resources []gardencorev1beta1.NamedResourceReference,
+) error {
+	shootNamespace := *config.Flux.Namespace
+	secretsToKeep := sets.Set[string]{}
+
+	secretResources := config.AdditionalSecretResources
+	if config.Source != nil && config.Source.SecretResourceName != nil {
+		secretResources = append(secretResources, fluxv1alpha1.AdditionalResource{
+			Name:       *config.Source.SecretResourceName,
+			TargetName: ptr.To(config.Source.Template.Spec.SecretRef.Name),
+		})
+	}
+	for _, resource := range secretResources {
+		name, err := copySecretToShoot(ctx, log, seedClient, shootClient, seedNamespace, shootNamespace, resources, resource)
+		if err != nil {
+			return fmt.Errorf("failed to copy secret: %w", err)
+		}
+		secretsToKeep.Insert(name)
+	}
+
+	// cleanup unreferenced secrets
+	secretList := &corev1.SecretList{}
+	if err := shootClient.List(ctx, secretList,
+		client.InNamespace(shootNamespace),
```

Comment:
If the namespace changes, we are doomed, right?

Reply:
yes, but I think we are doomed either way, we'd try to install flux again.... The correct thing would be to make this field immutable in a validating webhook, which is on the "v1" TODO list...

Reply:
Agreed, this is what I also had in mind. Just wanted to double-check.